### PR TITLE
Embedding LoRA A Fwd Kernel

### PR DESCRIPTION
**Pull request by srujan-thomdapu**
## Summary
Adds an XPU (SYCL) `embedding_lora_a_fwd` kernel for LoRA A embedding lookup with segment-based adapter selection. Integrated with PyTorch API for XPU.

## Changes
- Registers new `sgl_kernel.lora.embedding_lora_a_fwd` torch op for XPU
- Adds new SYCL kernel: `embedding_lora_a_fwd.cpp`
- Adds test: `test_embedding_lora_a_fwd.py` to validate against reference results (including multiple corner cases)
- Adds benchmarking: `bench_embedding_lora_a_fwd.py` to benchmark XPU kernel against the triton op (triton reference is taken from `frameworks.ai.pytorch.sglang` repo)

## Key Features
- Supports segment-based routing (`seg_indptr`, `weight_indices`)
- Handles multiple LoRA adapters with variable ranks
- Supports base vocab and extra token embeddings
- Zero-padding to `max_rank`

## Implementation Details
- 3D grid mapping directly assigns work across **segment / token-lane / rank-block** dimensions.
- Adapter selection is performed directly per segment using segment metadata.
- Output handling is performed in-kernel:
  - Writes gathered values from base embeddings or extra embeddings when valid.
  - Writes zeros for rank-tail padding, invalid tokens, and unsupported extra-token cases.

## Current Limitation
- The weight tensor layout is fixed as `(num_lora, max_rank, vocab_size)`, causing inherently strided reads across the rank dimension. This significantly impacts achievable memory bandwidth.

### Triton Op
- Uses `BLOCK_RANK` in the third grid dimension.
- Work-items are contiguous across the **rank dimension**.
- **Reads** are highly strided by `vocab_size` (large stride).
- **Writes** remain naturally coalesced.

### XPU Op
- Uses `BLOCK_TOK` in the third grid dimension.
- Work-items are contiguous across **token positions**.
- Since token IDs are arbitrary, reads are still irregular, but larger token batches may improve cache locality.
- **Writes** are strided by `max_rank`, making them less coalesced compared to Triton.

### Observation
- Benchmarking suggests that as token count and vocab size increase, the XPU op tends to outperform the Triton op due to improved cache behavior and better token-level scaling.

## Benchmarking
<details>
<summary><strong>Detailed Report</strong></summary>

| num_tokens | num_segments | num_adapters | max_rank | vocab_size | XPU FP16 (GB/s) | Triton FP16 (GB/s) | XPU BF16 (GB/s) | Triton BF16 (GB/s) | XPU FP32 (GB/s) | Triton FP32 (GB/s) |
|-----------:|-------------:|-------------:|---------:|-----------:|----------------:|-------------------:|----------------:|-------------------:|----------------:|-------------------:|
| 384   | 1  | 2 | 48  | 36864  | 0.42 | 0.77 | 0.48 | 0.78 | 0.87 | 1.46 |
| 512   | 1  | 4 | 64  | 40960  | 0.83 | 1.40 | 0.79 | 1.38 | 1.59 | 2.71 |
| 3072  | 2  | 4 | 77  | 49152  | 5.61 | 9.38 | 5.59 | 9.12 | 9.23 | 16.66 |
| 4608  | 8  | 2 | 96  | 57344  | 6.96 | 12.34 | 7.90 | 9.65 | 10.58 | 15.28 |
| 6144  | 4  | 4 | 123 | 65536  | 10.32 | 13.61 | 10.34 | 13.69 | 14.41 | 14.41 |
| 8192  | 16 | 8 | 128 | 73728  | 6.84 | 9.24 | 7.41 | 7.76 | 11.53 | 11.75 |
| 9216  | 32 | 4 | 175 | 81920  | 6.44 | 7.15 | 6.73 | 6.93 | 11.18 | 11.01 |
| 10240 | 8  | 4 | 192 | 98304  | 9.50 | 7.96 | 9.66 | 7.73 | 13.95 | 11.03 |
| 11264 | 32 | 2 | 221 | 65536  | 8.33 | 8.81 | 8.22 | 8.68 | 12.19 | 12.35 |
| 12288 | 64 | 8 | 256 | 86016  | 5.98 | 5.57 | 5.92 | 5.57 | 10.61 | 9.50 |
| 14336 | 16 | 4 | 320 | 98304  | 8.64 | 6.22 | 8.74 | 6.28 | 13.35 | 9.70 |
| 15872 | 64 | 2 | 365 | 110592 | 5.98 | 5.35 | 5.98 | 5.33 | 10.82 | 9.11 |
| 16384 | 32 | 4 | 448 | 122880 | 6.69 | 5.06 | 6.58 | 5.03 | 11.37 | 8.73 |
| 18432 | 64 | 2 | 512 | 114688 | 6.49 | 5.20 | 6.44 | 5.05 | 11.52 | 8.89 |
| 6144  | 2  | 4 | 69  | 53248  | 10.38 | 15.96 | 10.26 | 15.94 | 16.27 | 27.79 |
| 10240 | 4  | 2 | 144 | 69632  | 16.71 | 14.94 | 16.69 | 14.21 | 20.92 | 15.03 |
| 14336 | 32 | 2 | 192 | 86016  | 7.27 | 8.20 | 7.58 | 8.11 | 11.72 | 11.67 |
| 16384 | 16 | 4 | 576 | 98304  | 9.48 | 5.19 | 9.44 | 5.09 | 14.18 | 8.60 |
| 19456 | 64 | 4 | 640 | 122880 | 5.80 | 4.46 | 5.59 | 4.52 | 10.19 | 8.12 |
| 5050  | 1  | 2 | 288 | 120832 | 17.30 | 6.28 | 17.23 | 6.26 | 22.33 | 9.88 |
| 7000  | 1  | 1 | 95  | 118784 | 14.36 | 11.10 | 14.31 | 11.03 | 20.28 | 13.28 |
| 9500  | 2  | 4 | 230 | 116736 | 19.06 | 7.04 | 19.23 | 7.14 | 23.73 | 10.10 |
| 12000 | 4  | 2 | 704 | 122880 | 16.28 | 4.60 | 16.12 | 4.65 | 19.88 | 8.15 |
| 13824 | 8  | 4 | 460 | 110592 | 11.76 | 5.36 | 11.81 | 5.42 | 15.94 | 9.09 |
| 16384 | 16 | 8 | 384 | 98304  | 9.52 | 5.91 | 9.40 | 5.85 | 13.91 | 9.53 |
| 21504 | 32 | 2 | 731 | 121856 | 7.42 | 4.84 | 7.43 | 4.86 | 12.18 | 8.73 |

</details>

<details>
<summary><strong>Summary</strong></summary>

| Provider | Mean Bandwidth (GB/s) | Mean Time (ms) |
|----------|----------------------:|---------------:|
| XPU FP16    | 9.01  | 2.107 |
| XPU BF16    | 9.07  | 2.118 |
| XPU FP32    | 13.26 | 2.592 |
| Triton FP16 | 7.38  | 3.071 |
| Triton BF16 | 7.16  | 3.084 |
| Triton FP32 | 10.87 | 3.616 |

</details>

### Key Takeaway
The XPU implementation consistently improves as token count and workload size increase. In larger token-heavy cases, token-oriented execution shows better effective cache behavior and outperforms Triton despite non-coalesced writes.

## TODO / Follow-up Optimization
- Explore shared memory (local memory) based tiling as a follow-up optimization.
- Investigate reading weight tiles in a token-parallel manner to improve cache locality / partial read coalescing where possible.
- Use local memory as a temporary transpose buffer, followed by a barrier, to remap data and make output writes more coalesced across the rank dimension.
- Benchmark this approach against the current implementation to evaluate whether the added synchronization and local memory overhead provide a net bandwidth gain.

### CODE REVIEW

File: src/torch_extension_sycl.cc [hunk 1]
```diff
@@ -100,6 +99,7 @@ TORCH_LIBRARY_FRAGMENT(sgl_kernel, m) {
 
   m.def("merge_state_v2(Tensor v_a, Tensor s_a, Tensor v_b, Tensor s_b, Tensor! v_merged, Tensor! s_merged) -> ()");
   m.impl("merge_state_v2", torch::kXPU, &merge_state_v2);
+#ifdef BUILD_FMHA
   /*
    * From cutlass attention
    */
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/torch_extension_sycl.cc:105]: Removed the guard.

File: src/sycl/EmbeddingLoraAFwd.cpp [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+    // Some launched workgroups are padding because token_idx runs up to multiples of BLOCK_TOK
+    if (token_idx >= seg_len) {
+      return;
+    }
+
+    // Some local threads are padding because rank_idx runs up to max_rank
+    if (rank_idx >= rank) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/sycl/EmbeddingLoraAFwd.cpp:140]: output is explicitly zeroed at C++ entrypoint before launching the kernel

File: src/sycl/EmbeddingLoraAFwd.cpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+  }
+
+  TORCH_CHECK(input_ids.dim() == 1, "input_ids must be a 1D tensor");
+  TORCH_CHECK(weights.dim() == 3, "weights must be a 3D tensor");
+  TORCH_CHECK(weights.size(2) == vocab_size, "weights' vocab_size dimension must match the provided vocab_size");
+  TORCH_CHECK(seg_indptr.dim() == 1, "seg_indptr must be a 1D tensor");
+  TORCH_CHECK(weight_indices.dim() == 1, "weight_indices must be a 1D tensor");
+  TORCH_CHECK(lora_ranks.dim() == 1, "lora_ranks must be a 1D tensor");
+  TORCH_CHECK(output.dim() == 2, "output must be a 2D tensor");
+
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/sycl/EmbeddingLoraAFwd.cpp:247]: Since the integer values of seg_indptr, weight_indices and lora_ranks (Not using seg_lens tensor in the updated kernel code as it is optional) are not too large to become UB after casting int64_t to int32. However, the updated code will safely cast to int32. As it is a trusted API behind pytorch wrappers, we likely don’t need expensive runtime bounds checks.

File: src/sycl/EmbeddingLoraAFwd.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    // Some local threads are padding because rank_idx runs up to max_rank
+    if (rank_idx >= rank) {
+      return;
+    }
+
+    const int token_pos = seg_start + token_idx;
+
+    const int64_t tok = input_ids[token_pos];
+    DType* out_row = output + static_cast<int64_t>(token_pos) * max_rank;
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/sycl/EmbeddingLoraAFwd.cpp:145]: seg_lens tensor is ignored as it is optional in the updated code.

File: src/sycl/EmbeddingLoraAFwd.cpp [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    torch::Tensor& output,           // [num_tokens, max_rank]
+    const torch::Tensor& input_ids,  // [num_tokens,]
+    const torch::Tensor& weights,    // [num_loras, max_rank, vocab_size]
+    const int64_t vocab_size,
+    const torch::Tensor& seg_indptr,                       // [num_segments + 1,]
+    const torch::Tensor& weight_indices,                   // [num_segments,]
+    const torch::Tensor& lora_ranks,                       // [num_loras,]
+    const std::optional<torch::Tensor>& extra_embeddings,  // [num_loras, num_extra_tokens, max_rank]
+    const std::optional<torch::Tensor>&
+        seg_lens  // [num_segments,] optional; currently unused, reserved for future per-segment optimizations
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/sycl/EmbeddingLoraAFwd.cpp:228]: seg_lens tensor is ignored as it is optional in the updated code. Hence, I'm not changing this comment.

File: src/sycl/EmbeddingLoraAFwd.cpp [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    TORCH_CHECK(extra_embeddings->size(0) == num_loras_i64, "extra_embeddings.size(0) must equal weights.size(0)");
+    TORCH_CHECK(
+        extra_embeddings->size(2) == max_rank_i64, "extra_embeddings.size(2) must equal max_rank (weights.size(1))");
+    TORCH_CHECK(
+        extra_embeddings->scalar_type() == weights.scalar_type(), "extra_embeddings dtype must match weights dtype");
+  }
+
+  TORCH_CHECK(
+      lora_ranks.min().item<int>() >= 0 && lora_ranks.max().item<int>() <= max_rank_i64,
+      "All values in lora_ranks must be within the range [0, max_rank]");
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply src/sycl/EmbeddingLoraAFwd.cpp]: Applied suggested optimization steps to avoid expensive runtime bound checks.

File: benchmark/bench_embedding_lora_a_fwd.py [hunk 7]
```diff
@@ -0,0 +1,871 @@
...
+
+    # Load the token ID
+    token_id = tl.load(input_ids + seg_start + token_idx)
+
+    # Process in chunks of BLOCK_RANK dimensions
+    num_blocks = tl.cdiv(rank_val, BLOCK_RANK)
+
+    for block_id in range(num_blocks):
+        rank_offset = tl.arange(0, BLOCK_RANK) + block_id * BLOCK_RANK
+        rank_mask = rank_offset < rank_val
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply benchmark/bench_embedding_lora_a_fwd.py:336]: This is a reference/baseline Triton kernel taken from an external repo(sglang) and is intentionally kept as-is for comparison purposes. The production kernel is the XPU C++ implementation; the Triton version exists only to provide a performance baseline. No changes planned.

File: benchmark/bench_embedding_lora_a_fwd.py [hunk 8]
```diff
@@ -0,0 +1,871 @@
...
+    # 	rank_sum=rank_sum,
+    # )
+    total_bytes = _estimate_bytes_triton(
+        num_tokens=case["num_tokens"],
+        num_segments=num_segments,
+        max_rank=case["max_rank"],
+        max_len=max_len,
+        elem_size=elem_size,
+        rank_sum=rank_sum,
+    )
```
[the remark this answers is not in the document]
Comment (srujan-thomdapu) [reply benchmark/bench_embedding_lora_a_fwd.py]: Renamed it as per the suggestion

**Review by kareemshaik80 — COMMENTED** [review]
LGTM

**Review by kareemshaik80 — APPROVED** [review]
